rescript/get_unite_data.py

```diff
@@ -0,0 +1,167 @@
+    query_data = requests.get(base_url + doi).json()
+    # Updates can be made to files in a DOI, so on the advice of the devs,
+    # only return the last (newest) file with this -1  vv
+    URL = query_data['data'][0]['attributes']['media'][-1]['url']
+    return URL
+
+# Testing
+_unite_get_url(doi = '10.15156/BIO/2938079')
+
+# with tempfile.TemporaryDirectory() as tmp_dir:
+tmp_dir = tempfile.mkdtemp()
+
+
+def _unite_get_tgz(url, download_path):
+    print('Downloading ' + url)
+    response = requests.get(url, stream=True)
+    if response.status_code != 200:
+        raise ValueError("Failed to download the file from " + url)
+    # Save .tgz file
+    tar_file_path = os.path.join(download_path, 'unitefile.tar.gz')
+    with open(tar_file_path, 'wb') as f:
+        f.write(response.content)
+    # Extract only the 'developer' subdirectory
+    with tarfile.open(tar_file_path, 'r:gz') as tar:
+        # Ensure that 'developer' exists in the tar file
+        members = [member for member in tar.getmembers() if member.name.startswith('developer')]
+        if not members:
+            raise ValueError("No 'developer' subdirectory found in the .tar.gz file.")
+        for member in members:
+            member.name = os.path.basename(member.name) # Strip the 'developer' prefix
+            tar.extract(member, path=download_path)
+    
+    return download_path
+
+# Test it by downloading this file
+_unite_get_tgz(_unite_get_url(doi = '10.15156/BIO/2938079'), tmp_dir)
+
+
+def get_unite_data(version, taxon_group, singletons=False):
+    doi = _unite_get_doi(version, taxon_group, singletons)
+    url = _unite_get_url(doi)
+    # Eventual output
+    results = {'sequences': [], 'taxonomy': []}
+    # open temp output files, somehow?
+    with tempfile.TemporaryDirectory() as tmpdirname:
+        # tmp_dir = tempfile.mkdtemp()
+        print('Temporary directory:', tmpdirname)
+        _unite_get_tgz(url, download_path=tmpdirname)
+        # Find and import artifacts based on suffix
+        for root, dirs, files in os.walk(tmpdirname):
+            for file in files:
+                print(results)
+                if file.endswith('.fasta'):
+                    fasta_file_name = os.path.join(root, file)
+                    print('found fasta: ' + fasta_file_name)
+                    results['sequences'].append(qiime2.Artifact.import_data('FeatureData[Sequence]', fasta_file_name))
+                elif file.endswith('.txt'):
+                    txt_file_name = os.path.join(root, file)
+                    print('found txt: ' + txt_file_name)
+                    results['taxonomy'].append(qiime2.Artifact.import_data('FeatureData[Taxonomy]', txt_file_name))
+    return results
+
+# Testing
+example_output = get_unite_data(version='9.0', taxon_group='fungi')
+
+example_output['sequences'][1]
+
+
+# Unzip and import at the same time...
+# This function does not use _unite_get_tgz()
+def get_unite_data2(version, taxon_group, singletons=False):
```

Comment:
I think I generally prefer the `get_unite_data` method over this... because it is easier to test separate blocks of code rather than one big function. Unless there is some efficiency to this code I am overlooking.

Reply:
I had duplicate code that first loops over files in .tgz
then loops over the output files to import them.

I can keep them seperate and try to clean up internal organization.

I guess it's OK to duplicate code if the functions are different?

Reply:
don't duplicate the code... just write a small function that you call each time you loop over one of the clustering levels